web/src/main/java/org/springframework/security/web/server/util/matcher/IpAddressServerWebExchangeMatcher.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.web.server.util.matcher;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+import reactor.core.publisher.Mono;
+
+import org.springframework.util.Assert;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.util.StringUtils;
+
+/**
+ * Matches a request based on IP Address or subnet mask matching against the remote
+ * address.
+ *
+ * @author Guirong Hu
+ * @since 5.7
+ */
+public class IpAddressServerWebExchangeMatcher implements ServerWebExchangeMatcher {
+
+	private final int nMaskBits;
+
+	private final InetAddress requiredAddress;
```

Comment:
https://github.com/spring-projects/spring-security/blob/0541341201b408a4cb13ec846c6dc4fcd5600650/web/src/main/java/org/springframework/security/web/util/matcher/IpAddressMatcher.java#L64-L66

I thought about doing this before, but `IpAddressMatcher` depends on `HttpServletRequest`. When the user's web application type is `reactive` instead of `servlet`, will there be a dependency error? 

For example, the user only added the `spring-boot-starter-webflux` dependency to the project, but did not add the `spring-boot-starter-web` dependency,the program will fail to compile.

If you think there is no problem, then I will modify it according to your suggestion.

Reply:
[Maybe we can just like at the beginning](https://github.com/spring-projects/spring-security/pull/10007#issue-678551908), extracted the `matches(String)` method from `IpAddressMatcher` into a new class for reuse.

For example:

* Create a new class `InetAddressMatcher`

```java
public boolean matches(String address) {
	......
}
```

* The original `IpAddressMatcher` was modified to:

```java
public final class IpAddressMatcher implements RequestMatcher {

	private final InetAddressMatcher inetAddressMatcher;

	......

	@Override
	public boolean matches(HttpServletRequest request) {
		return matches(request.getRemoteAddr());
	}

	public boolean matches(String address) {
		return this.inetAddressMatcher.matches(address);
	}

        ......
}
```

* Reuse `matches(String)` method in `IpAddressServerWebExchangeMatcher`:

```java
public class IpAddressServerWebExchangeMatcher implements ServerWebExchangeMatcher {

	private final InetAddressMatcher inetAddressMatcher;

	......

	@Override
	public Mono<MatchResult> matches(ServerWebExchange exchange) {
		return Mono.justOrEmpty(exchange.getRequest().getRemoteAddress())
				.map((remoteAddress) -> remoteAddress.getAddress().getHostAddress())
				.map(this.inetAddressMatcher::matches)
				.flatMap((matches) -> matches ? MatchResult.match() : MatchResult.notMatch())
				.switchIfEmpty(MatchResult.notMatch());
	}
  
        ......
}
```

`InetAddressMatcher` is equivalent to a tool class, you can use it anywhere instead of depends on `HttpServletRequest`.

Reply:
Thanks for the clarification, @terminux. I was fairly certain it will still work, because both of these classes are in `spring-security-web` (meaning no dependency issues). But just to make sure, I backported your commit with my suggestion to the 5.7.x branch and tested it with a real application. It does work fine. And that's because we aren't flowing through the method that accepts an `HttpServletRequest`.

So I'd prefer to try this suggestion first. If there's something we're not thinking of that causes issues for folks, we can always move towards your suggestion. I'm just hoping to avoid introducing a new public API for this if at all possible.

Does that sound OK?

Reply:
Ok, done.

Thank you very much for verifying my question, @sjohnr.